### DISCUSSION TITLE
Is there currently an API to do random sampling in a closed arbitrary mesh

### DISCUSSION THREAD

**Discussion by Liu-xiandong** [pyvista/pyvista#2923] [Q&A]
### Describe the feature you would like to be added.

For any closed grid, to support various random sampling methods.

As shown in the image below, in a cube：
![image](https://user-images.githubusercontent.com/85323580/176653426-f4f958f0-e74f-4e69-b05e-c8713bf76074.png)

The core key is to support all closed meshes.

### Pseudocode or Screenshots

```python 
cube = pv.Cube(bounds=(-10,2000,-10,200,-10,2500))
cube.sampling(npoints = 100, method = "random")
 ```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by banesullivan** [comment]
There is no direct API for sampling points within bounds but it is pretty easy to do with numpy:

```py
import pyvista as pv
import numpy as np

pv.global_theme.hidden_line_removal = False  

cube = pv.Wavelet(extent=(-10, 20, -10, 20, -10, 25))

# Perform the random sampling of points within bounds of mesh
n = 100
pts = np.column_stack([np.random.uniform(low=a, high=b, size=n) for a, b in np.reshape(cube.bounds, (3,2))])
points = pv.PolyData(pts)

# Sample the data values of the mesh to those points
sampled = points.sample(cube)

# plot
p = pv.Plotter()
p.add_mesh(cube, style='wireframe' )
p.add_mesh(sampled, point_size=10, render_points_as_spheres=True)
p.show()
```
![Screen Shot 2022-06-30 at 10 43 11 AM](https://user-images.githubusercontent.com/22067021/176732779-46d564d1-a0f1-427a-ab49-04e450ef1e76.png)

**Comment by MatthewFlamm** [comment]
This picture comes from https://github.com/pyvista/pyvista/discussions/2703#discussioncomment-2828140 where a method is given when the structure of the cells is known and is nice enough to be able to do this deterministically.

For an unstructured mesh, it will be more difficult.  I proposed a random algorithm in https://github.com/pyvista/pyvista/discussions/2703#discussioncomment-2828184 which will work for any cell type at the potential downside of it being slow for highly skewed meshes.  This will be possible in PyVista once https://github.com/pyvista/pyvista/pull/2905 is merged and is included in a release (hopefully 0.35.0).

**Comment by MatthewFlamm** [comment]
I have published https://github.com/MatthewFlamm/pyransame which aims to solve this problem.  Currently, it only works for 2D surface meshes that are `PolyData` and can be triangulated, and not 3D meshes, e.g. `UnstructuredGrid`. Getting this to work for 3D meshes, or at least some subset of 3D meshes, is next on the todo list for this package.
